File: SudokuTs/src/Puzzle.ts
import * as Collections from 'typescript-collections';
import { PuzzleSize, PuzzleSizeUtil } from './PuzzleSize';

export type cellType = string | null;

export class Puzzle {
    public puzzleSize: PuzzleSize = PuzzleSize.Undefined;
    public grid: Collections.Dictionary<string, cellType>
        = new Collections.Dictionary<string, cellType>();

    public readonly NullChar = '.';

    /**
     * Ctor for a sudoku Puzzle.
     *
     * @param definition
     * null, which just initializes an empty Puzzle,
     * or a number representing the size of the sudoku Puzzle,
     * or another Puzzle to copy
     */
    constructor(definition: any = null) {

        // is it anything
        if (definition !== null) {
            // is it the size
            if (typeof definition === 'number') {
                // it's the order of the sudoku board
                if (!PuzzleSizeUtil.isValid(definition)) {
                    throw new Error(`${definition} is an invalid board order.`);
                }

                this.puzzleSize = PuzzleSizeUtil.toPuzzleSize(definition as number);
                this.createGrid();

            } else if (definition instanceof Puzzle) {
                this.puzzleSize = definition.puzzleSize;
                this.createGrid();
                this.loadPuzzle(definition.toString());
            }
        }
    }

    public createGrid() {
        for (let rIndex: number = 1; rIndex <= this.puzzleSize; rIndex++) {
            for (let cIndex: number = 1; cIndex <= this.puzzleSize; cIndex++) {
                this.grid.setValue(`r${rIndex}c${cIndex}`, null);
            }
        }
    }

    /**
     * Takes a string or array of strings representation of a puzzle and loads it into a puzzle grid.
     *
     * @param definition
     *  The string to load the sudoku board from,
     *  or a string[] representing the rows to parse the sudoku board from,
     *
     * @error If the length of definition isn't valid.
     */
    public loadPuzzle(definition: any) {

        if (typeof definition === 'string') {
            // assume it's a compact string of numeric values
            const cellCount: number = definition.length;

            const size = this.inferPuzzleSizeByCellCount(cellCount);

            if (size !== this.puzzleSize) {
                this.puzzleSize = size;
                this.createGrid();
            }

            for (let rIndex = 0; rIndex < this.puzzleSize; rIndex++) {
                const rowString: string = definition.slice(rIndex * this.puzzleSize, (rIndex + 1) * this.puzzleSize);
                this.parseRow(rowString, rIndex);
            }
        } else if (Array.isArray(definition) && typeof definition[0] === 'string') {
            // assume it's an array of rows of strings
            const definitionArray: string[] = definition as string[];

            const size = this.inferPuzzleSizeByRowLength(definitionArray.length);

            if (size !== this.puzzleSize) {
                this.puzzleSize = size;
                this.createGrid();
            }

            for (let rIndex = 0; rIndex < this.puzzleSize; rIndex++) {
                const rowString: string = definitionArray[rIndex];
                this.parseRow(rowString, rIndex);
            }
        } else {
            throw new Error(`${typeof definition} is not supported.`);
        }
    }

    /**
     * Evaluates cellCount to determine if it is a valid number of cells for a Sudoku board.
     *
     * @param cellCount the count of board cells.
     *
     * @returns true is valid, false otherwise.
     */
    public parseRow(rowString: string, r: number) {
        const regexp = /^[a-g1-9]$/i;   // supports hex+'g'

        for (let c = 0; c < this.puzzleSize; c++) {
            const cellString: string = rowString.slice(c, c + 1);
            const isCellValue = regexp.test(cellString);

            const gridIndex = `r${r + 1}c${c + 1}`;
            if (! isCellValue) {
                this.grid.setValue(gridIndex, null);
            } else {
                this.grid.setValue(gridIndex, cellString);
            }
        }
    }

    /**
     * Converts a puzzle to a string representation.
     */
    public toString(): string {
        let rv: string = '';

        for (let rIndex: number = 1; rIndex <= this.puzzleSize; rIndex++) {
            for (let cIndex: number = 1; cIndex <= this.puzzleSize; cIndex++) {
                rv += this.convertToChar(this.grid.getValue(`r${rIndex}c${cIndex}`));
            }
        }

        return rv;
    }

    /**
     * Converts a Puzzle cell's value to a character.
     *
     * @param cellValue the value of a Puzzle cell
     */
    public convertToChar(cellValue: cellType | undefined): string {
        if (cellValue) {
            return cellValue;
        }
        return this.NullChar;
    }

    /**
     * Gets the PuzzleSize for total cell count in a puzzle grid.
     *
     * @param cellCount: The total cell count in a puzzle grid.
     */
    public inferPuzzleSizeByCellCount(cellCount: number, throwIfUndefined: boolean = true): PuzzleSize {
        const size = Math.sqrt(cellCount);

        if (!PuzzleSizeUtil.isValid(size) && throwIfUndefined) {
            throw new Error(`${cellCount} is an invalid cell count.`);
        }

        return PuzzleSizeUtil.toPuzzleSize(size);
    }

    /**
     * Gets the PuzzleSize from the length of a row in a puzzle grid.
     *
     * @param size: The length of a row in a puzzle grid.
     */
    public inferPuzzleSizeByRowLength(size: number, throwIfUndefined: boolean = true): PuzzleSize {

        if (!PuzzleSizeUtil.isValid(size) && throwIfUndefined) {
            throw new Error(`${size} is an invalid cell count.`);
        }

        return PuzzleSizeUtil.toPuzzleSize(size);
    }

    /**
     * Evaluates boardCellCount to determine if it is a valid number of cells for a Sudoku board.
     *
     * @param boardCellCount the count of board cells.
     *
     * @returns true is valid, false otherwise.
     */
    public isValidCellCount(boardCellCount: number): boolean {
        const validCellCounts = new Set<number>();
        [PuzzleSize.FourByFour, PuzzleSize.NineByNine, PuzzleSize.SixteenBySixteen].forEach(element => {
            validCellCounts.add(element * element);
        });

        return validCellCounts.has(boardCellCount);
    }
}

File: SudokuSharp/Engine/Puzzle.cs
using System.Runtime.CompilerServices;

[assembly: InternalsVisibleTo(@"SudokuSharp.Engine.Tests")]

namespace SudokuSharp.Engine
{
    using System;
    using System.Collections.Generic;
    using System.ComponentModel;
    using System.Linq;
    using System.Text;
    using Annotations;

    /// <summary>
    /// A simple implementation of a Sudoku puzzle!
    /// </summary>
    public sealed class Puzzle : INotifyPropertyChanged
    {
        private byte?[,] _grid; // only access through the wrapper property this[]!
        private PuzzleSize _puzzleSize = PuzzleSize.Undefined;

        /// <summary>
        ///     Default ctor.
        /// </summary>
        public Puzzle() : this(PuzzleSize.Undefined)
        {
            // Must force grid creation since default size is Undefined.
            CreateGrid(PuzzleSize.Undefined);
        }

        /// <summary>
        ///     Public ctor.
        /// </summary>
        /// <param name="puzzleSize">This may not be needed, could be inferred when puzzle is loaded.</param>
        public Puzzle(PuzzleSize puzzleSize)
        {
            Size = puzzleSize;
        }

        private Puzzle(Puzzle other)
        {
            // no null check because it is private.

            LoadPuzzle(other.ToString());
        }

        /// <summary>
        ///     Get the number of cells in the Puzzle
        /// </summary>
        public int CellCount => _grid.Length;

        /// <summary>
        ///     Gets the size of the Puzzle.
        /// </summary>
        public PuzzleSize Size
        {
            get => _puzzleSize;
            private set
            {
                if (value != _puzzleSize)
                {
                    _puzzleSize = value;
                    CreateGrid(_puzzleSize);
                    NotifyOfPropertyChanged();
                }
            }
        }

        /// <summary>
        ///     Access the grid via puzzle coordinates
        /// </summary>
        /// <param name="puzzleCoordinate">The coordinate to access</param>
        /// <returns>The cell value at that coordinate if set, null otherwise</returns>
        public byte? this[PuzzleCoordinate puzzleCoordinate]
        {
            get => this[puzzleCoordinate.Row, puzzleCoordinate.Col];
            set => this[puzzleCoordinate.Row, puzzleCoordinate.Col] = value;
        }

        /// <summary>
        ///     Internal property to protect grid for proper status management
        /// </summary>
        /// <param name="row">The row coordinate to access</param>
        /// <param name="column">The column coordinate to access</param>
        /// <returns>The cell value at that coordinate if set, null otherwise</returns>
        internal byte? this[int row, int column]
        {
            get => _grid[row, column];
            set
            {
                var currentValue = _grid[row, column];
                if (!value.Equals(currentValue))
                {
                    _grid[row, column] = value;
                    NotifyOfPropertyChanged();
                }
            }
        }

        /// <summary>
        ///     Clone a deep copy of this puzzle.
        /// </summary>
        /// <returns>The cloned puzzle.</returns>
        public Puzzle Clone()
        {
            return new Puzzle(this);
        }

        /// <summary>
        /// Gets all the coordinate keys of the puzzle.
        /// </summary>
        /// <returns>All the coordinates of the puzzle.</returns>
        public IEnumerable<PuzzleCoordinate> GetKeys()
        {
            return from row in Enumerable.Range(0, Size.ToInt32())
                from col in Enumerable.Range(0, Size.ToInt32())
                select new PuzzleCoordinate(row, col);
        }

        /// <summary>
        ///     Gets the PuzzleSize for total cell count in a puzzle grid.
        /// </summary>
        /// <param name="cellCount">The total cell count in a puzzle grid</param>
        /// <param name="throwIfUndefined">if true, throw, else return Undefined - defaults to true</param>
        /// <returns>the PuzzleSize</returns>
        public PuzzleSize InferPuzzleSizeByCellCount(int cellCount, bool throwIfUndefined = true)
        {
            var isValid = PuzzleSizeExtensions.ValidPuzzleSizes
                .Select(v => v.ToInt32() * v.ToInt32())
                .Any(v => v == cellCount);

            if (!isValid && throwIfUndefined)
            {
                throw new ArgumentOutOfRangeException(nameof(cellCount), cellCount, "Not a known grid cell count.");
            }

            return PuzzleSizeExtensions.ToPuzzleSize(Convert.ToInt32(Math.Sqrt(cellCount)));
        }

        /// <summary>
        ///     Takes a string representation of a puzzle and loads it into a puzzle.
        /// </summary>
        /// <param name="puzzleText">A text string to try to load into the puzzle.</param>
        /// <param name="charToCell">char to byte mapping dictionary</param>
        /// <returns>true if loaded, false if not</returns>
        public bool LoadPuzzle(string puzzleText, IDictionary<char, byte?> charToCell = null)
        {
            var rv = false;

            // infer puzzleSize from dimensions
            Size = InferPuzzleSizeByCellCount(puzzleText.Length);
            charToCell ??= Size.ValidCellValues()
                .ToDictionary(v => v < 16 ? $"{v:x}"[0] : 'g', v => (byte?)v);

            var maxCellValue = Size.ToInt32();
            var gridArea = maxCellValue * maxCellValue;

            if (puzzleText.Length == gridArea)
            {
                var rowNum = 0;
                var colNum = 0;
                foreach (var cellChar in puzzleText.ToLowerInvariant())
                {
                    var coord = new PuzzleCoordinate(rowNum, colNum);

                    if (charToCell.ContainsKey(cellChar))
                    {
                        this[coord] = charToCell[cellChar];
                    }
                    else
                    {
                        this[coord] = null;
                    }

                    colNum++;
                    if (colNum >= maxCellValue)
                    {
                        colNum = 0;
                        rowNum++;
                    }

                    rv = true;
                }
            }

            return rv;
        }

        /// <summary>
        ///     Takes an array of strings representation of a puzzle and loads it into a puzzle.
        /// </summary>
        /// <param name="puzzleStrings">An array of strings to try to load into the puzzle.</param>
        /// <returns>true if loaded, false if not</returns>
        public bool LoadPuzzle(string[] puzzleStrings)
        {
            var sb = new StringBuilder();
            foreach (var row in puzzleStrings)
            {
                sb.Append(row);
            }

            return LoadPuzzle(sb.ToString());
        }

        /// <summary>
        ///     Dumps this puzzle as a string.
        /// </summary>
        /// <param name="nullChar"></param>
        /// <param name="newLinePerRow">If true, adds newline characters for each row string</param>
        /// <param name="upperCase">If true, will return all hex-plus characters in uppercase</param>
        /// <returns>This NullableByte puzzle as a string</returns>
        public string ToString(char nullChar = '.', bool newLinePerRow = false, bool upperCase = false)
        {
            var rv = new StringBuilder();
            foreach (var row in this.ByRow())
            {
                if (newLinePerRow && rv.Length != 0)
                {
                    rv.AppendLine();
                }

                foreach (var cell in row)
                {
                    var convertedChar = ConvertToChar(cell.Value, nullChar);
                    convertedChar = upperCase ? convertedChar.ToUpperInvariant() : convertedChar;
                    rv.Append(convertedChar);
                }
            }

            return rv.ToString();
        }

        /// <summary>
        /// Used to convert a cell value to a character (single character string).
        /// </summary>
        /// <param name="cell"></param>
        /// <param name="nullChar"></param>
        /// <returns></returns>
        internal string ConvertToChar(byte? cell, char nullChar = '.')
        {
            // just avoiding a lot of HasValue-Value checks
            var cellValue = cell ?? -1;

            if (cellValue > 0 && cellValue < 16)
            {
                return cellValue.ToString("x");
            }

            if (cellValue == 16)
            {
                return "g";
            }

            return nullChar.ToString();
        }


        /// <summary>
        /// Creates a new grid based upon puzzleSize's int equivalent.
        /// </summary>
        /// <param name="puzzleSize">The size of the puzzle grid to create.</param>
        private void CreateGrid(PuzzleSize puzzleSize)
        {
            var gridDim = puzzleSize != PuzzleSize.Undefined ? puzzleSize.ToInt32() : 0;
            _grid = new byte?[gridDim, gridDim];

            Size = puzzleSize;

            // ReSharper disable once UseNameofExpression
            NotifyOfPropertyChanged(@"Item");
        }

        #region INotifyPropertyChanged implementation

        public event PropertyChangedEventHandler PropertyChanged;

        /// <summary>
        /// Invokes PropertyChanged event.
        /// </summary>
        /// <param name="propertyName">The name of the property changed.</param>
        [NotifyPropertyChangedInvocator]
        private void NotifyOfPropertyChanged([CallerMemberName] string propertyName = null)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }

        #endregion
    }
}